cpp/tests/column/column_test.cu

```diff
@@ -403,4 +403,53 @@ TYPED_TEST(TypedColumnTest, ColumnViewConstructorWithMask)
   EXPECT_NE(original_view.null_mask(), copy_view.null_mask());
 }
 
```

Comment:
Test the range expectations, negative start, etc.

Reply:
I can't test that as it's a detail of the `lists::detail::copy_slice` function (`lists::detail::slice`) function, that is not exposed in any public API.

Reply:
I see that array_tests.[cpp, cu] have explicit calls like this `auto results      = cudf::strings::detail::copy_slice(strings_view, start, end);`.  So clearly the tests are testing `strings::detail::copy_slice`. So why can't we test `lists:detail::copy_slice`?

Reply:
Because the `slice` function is an implementation detail, and [testing behaviour, not implementation](
https://testing.googleblog.com/2013/08/testing-on-toilet-test-behavior-not.html?m=1) is considered good practice.

Reply:
So by "can't", you mean "don't think I should"?

Reply:
I guess I'm confused about why there needs to be a *behavior* that `CUDF_EXPECTS` something, but we are not allowed to test that behavior.  If we can expect a behavior, we should be able to test it.

Reply:
> So by "can't", you mean "don't think I should"?

Yes -- apologies, I should have been more clear and explained my reasoning better. 

> I guess I'm confused about why there needs to be a behavior that CUDF_EXPECTS something, but we are not allowed to test that behavior. If we can expect a behavior, we should be able to test it.

TL;DR: I think we should test that behaviour if (and only if) it is exposed via a public API.

Long answer:

At the moment, we have a "private" `detail::copy_slice` function, only a part of whose functionality is exercised by a "public" `cudf::column` constructor. We have tests for that part. It's true that now we have untested code in libcudf, but it's OK because that code can not be exercised in any way using the libcudf API. If/when it is exposed publicly (e.g., in a `cudf::copy_slice`API), _then_ it becomes necessary to test for that behaviour.

In my opinion, testing that behaviour while it is private is actively harmful; consider the following scenario:

Suppose we wrote a test for the range checking the behaviour of `detail::copy_slice` today. Suppose that tomorrow, we made a public `cudf::copy_slice`. Another test for the range checking behaviour of `cudf::copy_slice` would be redundant, and understandably, could be skipped. At that point, we would have a dangerous situation where public behaviour is tested in a test for a private function, the latter of which could be dropped or changed at any time.

Reply:
OK, thanks for clarifying. I understand now.